### PR TITLE
feat(ko): Init support for the ko builder

### PR DESCRIPTION
**Pull request by halvards**
Adds the ko builder as an option when running `skaffold init` in a directory containing Go code.

The ko builder will show up as an option for directories containing a `go.mod` file.

The feature can be disabled by adding the `--XXenableKoInit=false` flag.

Tracking: #7131

### CODE REVIEW

File: cmd/skaffold/app/cmd/init.go [hunk 1]
```diff
@@ -65,6 +66,7 @@ func NewCmdInit() *cobra.Command {
 			{Value: &enableNewInitFormat, Name: "XXenableNewInitFormat", DefValue: false, Usage: "", Hidden: true, IsEnum: true},
 			{Value: &enableJibInit, Name: "XXenableJibInit", DefValue: true, Usage: "", Hidden: true, IsEnum: true},
 			{Value: &enableJibGradleInit, Name: "XXenableJibGradleInit", DefValue: false, Usage: "", Hidden: true, IsEnum: true},
+			{Value: &enableKoInit, Name: "XXenableKoInit", DefValue: true, Usage: "", Hidden: true, IsEnum: true},
```
Comment (tejal29) [inline cmd/skaffold/app/cmd/init.go:69]: Let's keep the default value false to start with. For most alpha features, we start with flag value false and then  switch it to true.

File: docs-v2/content/en/docs/pipeline-stages/builders/ko.md [hunk 2]
```diff
@@ -254,30 +254,10 @@ Useful tips for existing `ko` users:
...
-    kubectl:
-      manifests:
-      - ./config/**
-  EOF
-  ```
-
-  Replace `./config/` with the path to your Kubernetes manifest files.
+- You can generate `skaffold.yaml` files by examining an existing code base,
+  using the command
+  [`skaffold init`]({{< relref "/docs/pipeline-stages/init" >}}).
```
Comment (tejal29) [inline docs-v2/content/en/docs/pipeline-stages/builders/ko.md:259]:
```suggestion
  [`skaffold init --xxEnabaleKo=true``]({{< relref "/docs/pipeline-stages/init" >}}).
```

File: docs-v2/content/en/docs/pipeline-stages/init.md [hunk 3]
```diff
@@ -17,7 +17,8 @@ and [deploy](#deploy-config-initialization) config.
 
 1. [Docker]({{<relref "/docs/pipeline-stages/builders/docker">}})
 2. [Jib]({{<relref "/docs/pipeline-stages/builders/jib">}})
-2. [Buildpacks]({{<relref "/docs/pipeline-stages/builders/buildpacks">}})
+3. [Ko]({{<relref "/docs/pipeline-stages/builders/ko">}})
+4. [Buildpacks]({{<relref "/docs/pipeline-stages/builders/buildpacks">}})
```
Comment (tejal29) [inline docs-v2/content/en/docs/pipeline-stages/init.md:21]: Ko is preferred over buildpacks right?
if an artifact with `go.mod` is present, init will configure ko artifact over buildpacks?

File: pkg/skaffold/initializer/analyze/builder.go [hunk 4]
```diff
@@ -86,6 +88,14 @@ func (a *builderAnalyzer) detectBuilders(ctx context.Context, path string, detec
...
 	}
 
+	if a.enableKoInit {
+		if koinit.Validate(path) {
+			results = append(results, koinit.ArtifactConfig{
+				File: path,
+			})
+		}
+	}
+
```
Comment (tejal29) [inline pkg/skaffold/initializer/analyze/builder.go:98]: Can you potentially have multiple `go.mod` nested inside subdirectories?
If not, code should return `searchSubDirectories = false`

**Review by tejal29 — COMMENTED** [review]
LGTM. 

TODO
- @tejal29 to try out skaffold init on a go project.

File: docs-v2/content/en/docs/pipeline-stages/init.md (hunk 3 above)
Comment (halvards) [reply docs-v2/content/en/docs/pipeline-stages/init.md:21]: If a user runs `skaffold init` and a `go.mod` file is present, the user will get a prompt to select Buildpacks or Ko (the list of builder options) is sorted alphabetically:

```shell
$ skaffold init
? Choose the builder to build image skaffold-ko  [Use arrows to move, type to filter]
> Buildpacks (go.mod)
  Ko (go.mod)
  None (image not built from these sources)
```

If the user adds the `--force` flag, then init will select the Ko builder in preference over Buildpacks:

```shell
$ skaffold init --force
Configuration skaffold.yaml was written
You can now run [skaffold build] to build the artifacts
or [skaffold run] to build and deploy
or [skaffold dev] to enter development mode, with auto-redeploy
```

```shell
$ cat skaffold.yaml
apiVersion: skaffold/v3alpha1
kind: Config
metadata:
  name: ko
build:
  artifacts:
  - image: skaffold-ko
    ko:
      dependencies:
        paths:
        - '**/*.go'
        - go.*
      dir: .
manifests:
  rawYaml:
  - k8s/example.yaml
```

File: pkg/skaffold/initializer/analyze/builder.go (hunk 4 above)
Comment (halvards) [reply pkg/skaffold/initializer/analyze/builder.go:98]: > Can you potentially have multiple `go.mod` nested inside subdirectories?

Yes, and that should result in multiple artifacts.

File: cmd/skaffold/app/cmd/init.go (hunk 1 above)
Comment (halvards) [reply cmd/skaffold/app/cmd/init.go]: I pushed another commit that sets the flag value to false by default.

File: docs-v2/content/en/docs/pipeline-stages/builders/ko.md (hunk 2 above)
Comment (halvards) [reply docs-v2/content/en/docs/pipeline-stages/builders/ko.md]: Updated docs in new commit.

**Comment by halvards** [comment]
Thanks for the review @tejal29 ! I pushed another commit with your suggested changes.

**Review by tejal29 — APPROVED** [review]
LGTM

**Pull request merged**